osu.Game/Screens/Edit/Compose/Components/LinedPositionSnapGrid.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Utils;
+using osuTK;
+
+namespace osu.Game.Screens.Edit.Compose.Components
+{
+    public abstract partial class LinedPositionSnapGrid : PositionSnapGrid
+    {
+        protected void GenerateGridLines(Vector2 step, Vector2 drawSize)
+        {
+            if (Precision.AlmostEquals(step, Vector2.Zero))
+                return;
+
+            int index = 0;
+
+            // Make lines the same width independent of display resolution.
+            float lineWidth = DrawWidth / ScreenSpaceDrawQuad.Width;
+            float rotation = MathHelper.RadiansToDegrees(MathF.Atan2(step.Y, step.X));
+
+            List<Box> generatedLines = new List<Box>();
+
+            while (true)
+            {
+                Vector2 currentPosition = StartPosition.Value + index++ * step;
+
+                if (!lineDefinitelyIntersectsBox(currentPosition, step.PerpendicularLeft, drawSize, out var p1, out var p2))
+                {
+                    if (!isMovingTowardsBox(currentPosition, step, drawSize))
+                        break;
```

Comment:
For a protected method it feels slightly strange that what this will do if given a vector that goes in the wrong direction relative to the origin, is give up once it is outside of expected range... I guess it's better than looping forever (and also that this is the termination condition even in the normal case) but it's a bit weird.

Is this gonna be used for non-rectangular grids too? I assume yes, in which case I think a harder assertion that the step component needs to have both coordinates positive wouldn't hurt.